File: packages/shared-utils/src/lib/money.ts
/**
 * Monetary helpers. FinanceHub stores money as an integer number of **minor
 * units** to avoid floating-point rounding errors. These helpers convert
 * to/from the display representation.
 *
 * The minor-unit exponent is *not* always 2: JPY has 0 (¥1000 is 1000 minor
 * units, not ¥10) and KWD has 3. It is read from ICU via `Intl` rather than
 * from a hand-maintained table.
 */

const DEFAULT_CURRENCY = 'USD';

/** ISO 4217 minor-unit factor for `currency` (JPY → 1, USD → 100, KWD → 1000). */
function minorUnitFactor(currency: string): number {
  let digits = 2;
  try {
    digits =
      new Intl.NumberFormat('en-US', {
        style: 'currency',
        currency,
      }).resolvedOptions().maximumFractionDigits ?? 2;
  } catch {
    // Not a well-formed currency code: fall back to the 2-digit majority
    // rather than blowing up a whole rendered table.
  }
  return 10 ** digits;
}

/** Converts minor units to a major-unit number (e.g. `4599` USD → `45.99`). */
export function minorToMajor(
  amountMinor: number,
  currency: string = DEFAULT_CURRENCY,
): number {
  return amountMinor / minorUnitFactor(currency);
}

/** Converts a major-unit amount to integer minor units. */
export function majorToMinor(
  amountMajor: number,
  currency: string = DEFAULT_CURRENCY,
): number {
  return Math.round(amountMajor * minorUnitFactor(currency));
}

/**
 * Formats an amount in minor units as a localized currency string.
 *
 * @param amountMinor Integer amount in minor units.
 * @param currency    ISO 4217 currency code (e.g. `USD`, `EUR`, `BRL`).
 * @param locale      BCP 47 locale tag; defaults to `en-US`.
 */
export function formatMoney(
  amountMinor: number,
  currency: string,
  locale = 'en-US',
): string {
  return new Intl.NumberFormat(locale, {
    style: 'currency',
    currency,
  }).format(minorToMajor(amountMinor, currency));
}

File: apps/mobile/lib/src/core/format/money.dart
import 'dart:math' as math;

import 'package:intl/intl.dart';

import '../config/env.dart';

/// Monetary helpers mirroring the backend/web contract: money is stored as an
/// integer number of **minor units** to avoid float rounding.
///
/// The exponent is not always 2 — JPY has 0 and KWD has 3 — so it is read from
/// ICU through `intl` instead of being hardcoded to cents.
class Money {
  const Money._();

  /// ISO 4217 minor-unit factor (JPY → 1, USD → 100, KWD → 1000).
  static int _factor(String currency) {
    final digits = NumberFormat.simpleCurrency(locale: 'en_US', name: currency)
            .decimalDigits ??
        2;
    return math.pow(10, digits).toInt();
  }

  /// Minor units → major-unit double (e.g. `4599` + `USD` → `45.99`).
  static double minorToMajor(int amountMinor, [String currency = 'USD']) =>
      amountMinor / _factor(currency);

  /// Major-unit amount → integer minor units.
  static int majorToMinor(num amountMajor, [String currency = 'USD']) =>
      (amountMajor * _factor(currency)).round();

  /// Formats minor units as a localized currency string
  /// (e.g. `4599` + `USD` → `$45.99`).
  static String format(
    int amountMinor,
    String currency, {
    String? locale,
  }) {
    final format = NumberFormat.simpleCurrency(
      locale: locale ?? Env.locale,
      name: currency,
    );
    return format.format(minorToMajor(amountMinor, currency));
  }

  /// A signed variant that always shows a leading `+`/`−` — handy for ledgers.
  static String formatSigned(
    int amountMinor,
    String currency, {
    String? locale,
  }) {
    final sign = amountMinor > 0 ? '+' : (amountMinor < 0 ? '−' : '');
    return '$sign${format(amountMinor.abs(), currency, locale: locale)}';
  }
}
